src/controller/JCMCommissioner.cpp

```diff
@@ -0,0 +1,430 @@
+    }
+    
+    if (result != JCMTrustVerificationResult::kSuccess)
+    {
+        // Handle error
+        ChipLogError(Controller, "JCM: Error in Trust Verification: %d", static_cast<int>(result));
+        OnTrustVerificationComplete(result);
+        return;
+    }
+
+    switch (mNextStage)
+    {
+        case chip::Controller::JCMTrustVerificationStage::kStarted:
+            mNextStage = JCMTrustVerificationStage::kVerifyingAdministratorEndpointAndFabricIndex;
+            VerifyAdministratorEndpointAndFabricIndex();
+            break;
+        case JCMTrustVerificationStage::kVerifyingAdministratorEndpointAndFabricIndex:
+            mNextStage = JCMTrustVerificationStage::kPerformingVendorIDVerificationProcedure;
+            PerformVendorIDVerificationProcedure();
+            break;
+        case JCMTrustVerificationStage::kPerformingVendorIDVerificationProcedure:
+            mNextStage = JCMTrustVerificationStage::kVerifyingNOCContainsAdministratorCAT;
+            VerifyNOCContainsAdministratorCAT();
+            break;
+        case JCMTrustVerificationStage::kVerifyingNOCContainsAdministratorCAT:
+            mNextStage = JCMTrustVerificationStage::kAskingUserForConsent;
+            AskUserForConsent();
+            break;
+        case JCMTrustVerificationStage::kAskingUserForConsent:
+            mNextStage = JCMTrustVerificationStage::kIdle;
+            OnTrustVerificationComplete(result);
+            break;
+        default:
+            ChipLogError(Controller, "JCM: Invalid stage: %d", static_cast<int>(mNextStage));
+            OnTrustVerificationComplete(JCMTrustVerificationResult::kInternalError);
+            break;
+    }
+}
+
+void JCMDeviceCommissioner::CleanupCommissioning(DeviceProxy * proxy, NodeId nodeId, const CompletionStatus & completionStatus)
+{
+    mNextStage = JCMTrustVerificationStage::kIdle;
+    mInfo.Cleanup();
+
+    DeviceCommissioner::CleanupCommissioning(proxy, nodeId, completionStatus);
+}
+
+/*
+ * JCMAutoCommissioner override implementation
+*/
+CHIP_ERROR JCMAutoCommissioner::SetCommissioningParameters(const CommissioningParameters & params)
+{
+    ReturnErrorOnFailure(AutoCommissioner::SetCommissioningParameters(params));
+
+    if (params.UseJCM().ValueOr(false)) {
+        auto extraReadPaths = params.GetExtraReadPaths();
+
+        mTempReadPaths.clear();
+        mTempReadPaths.reserve(extraReadPaths.size() + mExtraReadPaths.size());
+        mTempReadPaths.insert(mTempReadPaths.end(), extraReadPaths.begin(), extraReadPaths.end());
+        mTempReadPaths.insert(mTempReadPaths.end(), mExtraReadPaths.begin(), mExtraReadPaths.end());
+
+        // Set the extra read paths for JCM
+        mParams.SetExtraReadPaths(Span<app::AttributePathParams>(mTempReadPaths.data(), mTempReadPaths.size()));
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+void JCMAutoCommissioner::CleanupCommissioning()
+{
+    mTempReadPaths.clear();
+
+    AutoCommissioner::CleanupCommissioning();
```

Comment:
In AutoCommissioner::CleanupCommissioning() the memory allocated for mPAI and mDAC is freed. In JCMAutoCommissioner::CleanupCommissioning() we need to free up the memory allocated for mAdminICAC, mAdminNOC, mAdminRCAC. Or does it happen in a different place?

Reply:
adminICAC, adminNOC, and adminRCAC is freed in the overridden JCMDeviceCommissioner::CleanupCommissioning() methods call to the JCMTrustVerificationInfo::Cleanup method.

Reply:
@doru91 Verified it is being  called